src/kernel/src/main.rs

```diff
@@ -40,26 +43,14 @@ mod syscalls;
 mod sysctl;
 mod ucred;
 
-fn main() -> ExitCode {
+fn main() -> Result<(), KernelError> {
```

Comment:
Could you show some examples of the output? Not sure if this supports nested error.

Reply:
Error: Failed to open .kernel-debug: No such file or directory (os error 2)

Reply:
It seems that it can nest arbitrarily. VocalFan can confirm

Reply:
By the way, all those match statements all over the codebase could be refactored with #[from] instead of #[source]

Reply:
> Error: Failed to open .kernel-debug: No such file or directory (os error 2)

This one can't be tested to see if it is support nested error. You need something like this https://github.com/obhq/obliteration/issues/401#issue-1947775969.

Reply:
I tested it in Discord, we made it able to match issue 401's error, which is what the later Commits are, @ultimaweapon

Reply:
Yup. VocalFan tested that exact one and it worked

Reply:
So if I understand correctly, the issue is that with external errors we have to call ->source() to the nested one and the error! macro did that? If so, I could implement the Termination trait manually and use error!

Reply:
Correct. With `Error::source` it is guaranteed that we will be able to get the inner error from any errors.

Reply:
Should I also use #[from] to simplify all the other errors outside of main? @ultimaweapon. I think it would be more readable

Reply:
How we will handle this case?

```rust
fn foo() -> Result<(), FooError> {
    let file = File::open("abc")?;
    file.read(&mut buf)?;
}

#[derive(Debug, Error)]
enum FooError {
    Bar(#[from] std::io::Error)
}
```

Reply:
What's the issue here?

Reply:
I already looked at all the existing code and there are definitely some issues, but I don't quite see one here. One of the issues is that more that one variant contains the same error and another one is that a lot of variants contain more info than just the error. This could be fixed by .map_err, but I'm not quite sure if that's your style of code, that's up to you. There's also a lot of opportunities to use .ok_or on Option but again, not sure if you consider that better than a match statement. You tell me

Reply:
The problem with `#[from]` here is easy to accidentally return the wrong variant like my above example that a single variant will represent both error for `open` and `read`.

I'm fine with both `map_err` and `ok_or` as long as the call chain is not too long and it does not require cloning a heap allocated value.

Reply:
I see. I don't suppose we can do anything about that, other than just watch out.